### ISSUE TITLE
MIOpen Immediate Mode 

### ISSUE THREAD

**Issue by Delaunay** [ROCm/pytorch#433]
I have seen that MIOpen 2.0 supports [immediate mode][1].
I wondering if you guys have plans to support it ?

I am thinking it would be the default and then something like `torch.backend.miopen.benchmark/find = True` to enable the more lengthy find algorithm.

As currently the first batch can be very slow.

[1]:https://rocmsoftwareplatform.github.io/MIOpen/doc/html/find_and_immediate.html?highlight=immediate#immediate-mode-api

**Comment by iotamudelta** [comment]
We are evaluating it. It'll likely boil down to performance considerations for immediate mode vs traditional for us.

**Comment by Delaunay** [comment]
Well depending on the use case this might be important.
If you have an already known model that you want to train this feature is not really interesting.

But it might be a source of frustration if you are quickly prototyping models/debugging a pipeline.
One workflow our researchers follow is to try a few quick tests on CIFAR (few minutes to hours) before moving to ImageNet for the final paper.

To illustrate, you can find the time spent on the first batch below.
It might become significant when doing Neural Architecture Search on small datasets.

* MI50 initial compile time

| Model   | Batch size | first init (s) | second init (sec) |
|---------|------------|----------------|-------------------|
|resnet18 |  256       | 63             | 10                |
|resnet18 |  128       | 61             |  8                |
|resnet50 |  32        | 128            | 17                |
|resnet50 |  68        | 145            | 20                |


*  Nvidia initial compile time

| Model   | Batch size | first init (s) | second init (sec) |
|---------|------------|----------------|-------------------|
|resnet18 |  256       |  1.38          |   0.78            |
|resnet18 |  128       |  0.46          |   0.48            |
|resnet50 |  32        |  0.45          |   0.46            |
|resnet50 |  68        |  0.77          |   0.75            |

so I think having the ability to enable immediate mode would be useful.

**Comment by dagamayank** [comment]
@iotamudelta Can we reproduce the MI50 compilation times for 1st and 2nd inits? I am surprised that the 2nd invocation is also taking that long. That should be cached.

**Comment by qixiang109** [comment]
just wondering any update on this problem? @Delaunay your compile time is the same with mine. Very bad experience

**Comment by iotamudelta** [comment]
@qixiang109 we are working on this issue. I can confirm that switching to immediate mode for PyTorch works in our bringup. We are working on assuring the same level of performance is achieved in that setup before we can make the switch.

**Comment by qixiang109** [comment]
> @qixiang109 we are working on this issue. I can confirm that switching to immediate mode for PyTorch works in our bringup. We are working on assuring the same level of performance is achieved in that setup before we can make the switch.

thank you for your job that will be a great progress I believe. btw, will the immediate mode also works with tensorflow?

**Comment by atamazov** [comment]
@qixiang109 At MIOpen team, we are also working on performance improvements of existing "conventional" Find mode of the library. These improvements include accelerated Fast Find and Hybrid Find modes that use Immediate mode machinery internally. This way, any application (including TF) may get the faster startup times without need to change anything but set some environment variable.

Fast/Hybrid Find modes are implemented and currently under testing. Stay tuned.

**Comment by Delaunay** [comment]
@atamazov Is there investigation into doing in-memory compilation instead of generating files to be compiled and compile them ? Theano got a huge compile time boost by doing so, removing all the IO latency from the compilation.

**Comment by atamazov** [comment]
@Delaunay Yes, however I do not expect substantial acceleration from in-memory compilation, because the most of the time is spent not for IO but within the compiler. And there are some other ongoing developments targeted to improvements of startup time, e.g. precompiled kernels in MIOpen and improvements of the ROCm HCC compiler itself.

**Comment by Delaunay** [comment]
Good to know.
I am guessing the hybrid immediate mode, queue the compilation and returns the immediate kernel and then switch to the compiled kernel when ready, so with this kernels could be compiled in parallel.
This should improve the speed quite significantly indeed.

A while back, I also noticed that when I start pytorch in parallel for multi GPU training, each process will go through the compilation process independently, depending on the number of processor the machine has; it can make the compilation time much slower.
Given the nature of python I am not sure how much of this can be avoided.
I guess a lock file could be introduced per kernels to prevent MIOpen from starting the work on something that is already being compiled. It was a few months ago so maybe you have already solved it.

**Comment by qixiang109** [comment]
> @qixiang109 At MIOpen team, we are also working on performance improvements of existing "conventional" Find mode of the library. These improvements include accelerated Fast Find and Hybrid Find modes that use Immediate mode machinery internally. This way, any application (including TF) may get the faster startup times without need to change anything but set some environment variable.
> 
> Fast/Hybrid Find modes are implemented and currently under testing. Stay tuned.

eagerly looking forward of the new modes!

**Comment by atamazov** [comment]
@qixiang109 New modes available for testing at https://github.com/ROCmSoftwarePlatform/MIOpen/pull/162

**Comment by naromero77amd** [comment]
Closing this issue as this has already been implemented. miopen backends map into the cudnn backends.
https://pytorch.org/docs/stable/backends.html#module-torch.backends.cudnn

Immediate mode maps into:
```
torch.backends.cudnn.benchmark=false 
```
For find mode, set it to `true`

Note that miopen modes settings have changed recently upstream.

Now immediate mode requires:
```
torch.backends.cudnn.benchmark=false
torch.backends.cudnn.deterministic=true
```

More information can be found in this PR:
https://github.com/pytorch/pytorch/pull/145294
